api/Updated-Metrics.md

```diff
@@ -0,0 +1,406 @@
+Here we have shown binary values for a simple classification problem, but they could be floating point values from a regression, or even collections of classes and associated probabilities.
+
+## Basic Calls
+
+### Existing Syntax
+
+Our basic method is `group_summary()`
+
+```python
+>>> result = flm.group_summary(skm.accuracy_score,
+                               y_true, y_pred,
+                               sensitive_features=A_1)
+>>> print(result)
+{'overall': 0.4, 'by_group': {'B': 0.6536, 'C': 0.213}}
+>>> print(type(result))
+<class 'sklearn.utils.Bunch'>
+```
+The `Bunch` is an object which can be accessed in two ways - either as a dictionary - `result['overall']` -  or via properties named by the dictionary keys - `result.overall`.
+Note that the `by_group` key accesses another `Bunch`.
+
+We allow for sample weights (and other arguments which require slicing) via `indexed_params`, and passing through other arguments to the underlying metric function (in this case, `normalize`):
+```python
+>>> flm.group_summary(skm.accuracy_score,
+                      y_true, y_pred,
+                      sensitive_features=A_1,
+                      indexed_params=['sample_weight'],
+                      sample_weight=weights, normalize=False)
+{'overall': 20, 'by_group': {'B': 60, 'C': 21}}
+```
+
+We also provide some wrappers for common metrics from SciKit-Learn:
+```python
+>>> flm.accuracy_score_group_summary(y_true, y_pred, 
+                                     sensitive_features=A_1)
+{'overall': 0.4, 'by_group': {'B': 0.6536, 'C': 0.213}}
+```
+
+### Proposed Change
+
+We propose to introduce a new object, the `GroupedMetric` (name discussion below).
+Users will compute metrics by passing arguments into the constructor:
+```python
+>>> metrics = GroupedMetrics(skm.accuracy_score,
+                             y_true, y_pred,
+                             sensitive_features=A_1)
+<class 'GroupedMetric'>
+>>> metrics.overall
+          accuracy_score
+overall   0.4
+>>> metrics.by_group
+   accuracy_score
+B             0.6536
+C             0.213
+```
+Both properties are now Pandas DataFrames, with the column name set to the `__name__` property of the given metric function.
+The rows of the `by_group` property are set to the unique values of the `sensitive_feature=` argument.
+
+Any extra arguments for the metric function would be passed via a `params=` dictionary (and an `indexed_params=` list):
+```python
+>>> acc_params = { 'sample_weight': weights, 'normalize':False }
+>>> metrics = GroupedMetrics(skm.accuracy_score,
+                             y_true, y_pred,
+                             sensitive_features=A_1,
+                             indexed_params=['sample_weight'],
+                             params=acc_params)
+```
+We would _not_ provide the basic wrappers such as `accuracy_score_group_summary()`.
+
```

Comment:
At the level of the constructor, we should maybe follow the convention:
```python
metrics = GroupedMetrics(
    skm.accuracy_score, y_true, y_pred,
    sensitive_features=A_1,
    indexed_params={'sample_weight': weights},
    params={'normalize': False})
```

Reply:
So we sort of have two `**kwargs` dictionaries, instead of just one with a list of 'special' keys? I can see the sense in that.